lib/Service/AvatarService.php

```diff
@@ -0,0 +1,130 @@
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Service;
+
+use InvalidArgumentException;
+use OCA\Talk\Room;
+use OCP\Files\IAppData;
+use OCP\Files\NotFoundException;
+use OCP\Files\SimpleFS\InMemoryFile;
+use OCP\Files\SimpleFS\ISimpleFile;
+use OCP\IAvatarManager;
+use OCP\IL10N;
+use OCP\IUser;
+
+class AvatarService {
+	private IAppData $appData;
+	private IL10N $l;
+	private IAvatarManager $avatarManager;
+
+	public function __construct(
+		IAppData $appData,
+		IL10N $l,
+		IAvatarManager $avatarManager
+	) {
+		$this->appData = $appData;
+		$this->l = $l;
+		$this->avatarManager = $avatarManager;
+	}
+
+	public function setAvatar(Room $room, string $content): void {
+		$image = new \OC_Image();
+		$image->loadFromData($content);
+		$image->readExif($content);
+		$image->fixOrientation();
+		if (!($image->height() === $image->width())) {
+			throw new InvalidArgumentException($this->l->t('Avatar image is not square'));
+		}
+
+		if (!$image->valid()) {
+			throw new InvalidArgumentException($this->l->t('Invalid image'));
+		}
+
+		$mimeType = $image->mimeType();
+		$allowedMimeTypes = [
+			'image/jpeg',
+			'image/png',
+			'image/svg',
+		];
+		if (!in_array($mimeType, $allowedMimeTypes)) {
+			throw new InvalidArgumentException($this->l->t('Unknown filetype'));
+		}
+
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+		} catch (NotFoundException $e) {
+			$folder = $this->appData->newFolder('room-avatar');
+		}
+		$token = $room->getToken();
+		$folder->newFile($token, $image->data());
+	}
+
+	public function getAvatar(Room $room, ?IUser $user, bool $dark = false): ISimpleFile {
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+		} catch (NotFoundException $e) {
+			$folder = $this->appData->newFolder('room-avatar');
+		}
+		$token = $room->getToken();
+		if ($folder->fileExists($token . '.png')) {
+			$file = $folder->getFile($token . '.png');
+		} elseif ($folder->fileExists($token . '.jpeg')) {
+			$file = $folder->getFile($token . '.jpeg');
+		} else {
+			if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+				$users = json_decode($room->getName(), true);
+				foreach ($users as $participantId) {
+					if ($participantId !== $user->getUID()) {
+						$avatar = $this->avatarManager->getAvatar($participantId);
+						$file = $avatar->getFile(512);
+					}
+				}
```

Comment:
This should be first thing in the method? There is no use to do any of the above when it's the avatar of the other user anyway?

Reply:
This is the fallback, only is necessary if the room haven't an avatar.

This part I copied from here:
https://github.com/nextcloud/spreed/blob/master/lib/Search/ConversationSearch.php#L118-L138
And probability I will unify those codes after do more small adjusts in the code to cover the pending requirements.

Reply:
Unification of the code is not needed.
As per todo list you can remove all those iconClass checking there and just always return the icon URL with `.../avatar/token`

Reply:
I tried to remove the classes but I think won't be possible because the fallback images are in svg format and is used CSS to make with 22px in the center of circle with a border. This is a point but have other.

The search component use an `img` tag to display the avatar (maybe will be necessary replace by `NcAvatar`) and I don't know why but SVG files give an error when loading.

Reply:
Then keep the SVG stuff here, but the one-to-one part can be moved to the beginning of the method. No need to check the filesystem with it.